File: BinaryTrees/PreOrder/pre.cpp
#include <iostream>
#include <vector>
#include <stack>

using namespace std;

// Node Class
class Node {
public:
    int data;
    Node* left;
    Node* right;

    Node(int val) {
        data = val;
        left = nullptr;
        right = nullptr;
    }
};

/////////////////////////////////////////////////////////
// Recursive Preorder
/////////////////////////////////////////////////////////

class Solution1 {
public:
    void doPreOrder(Node* root, vector<int>& ans) {
        if (root == nullptr)
            return;

        ans.push_back(root->data);
        doPreOrder(root->left, ans);
        doPreOrder(root->right, ans);
    }

    vector<int> preOrder(Node* root) {
        vector<int> ans;
        doPreOrder(root, ans);
        return ans;
    }
};

/////////////////////////////////////////////////////////
// Iterative Preorder
/////////////////////////////////////////////////////////

class Solution2 {
public:
    vector<int> preOrder(Node* root) {

        vector<int> ans;

        if (root == nullptr)
            return ans;

        stack<Node*> st;
        st.push(root);

        while (!st.empty()) {

            Node* node = st.top();
            st.pop();

            ans.push_back(node->data);

            if (node->right)
                st.push(node->right);

            if (node->left)
                st.push(node->left);
        }

        return ans;
    }
};

File: BinaryTrees/PreOrder/pre.py
'''
# Node Class:
class Node:
    def __init__(self,val):
        self.data = val
        self.left = None
        self.right = None
'''

class Solution1:
    def doPreOrder(self,root,ans):
        if not root: return
        ans.append(root.data)
        self.doPreOrder(root.left,ans)
        self.doPreOrder(root.right,ans)
    def preOrder(self, root):
        ans = []
        self.doPreOrder(root,ans)
        return ans


class Solution2:
    def preOrder(self, root):
        ans = []
        st = [root]
        while len(st):
            node = st[-1]
            ans.append(node.data)
            st.pop()
            if node.right: st.append(node.right)
            if node.left: st.append(node.left)
        return ans
